### ISSUE TITLE
`indent`: wrong behavior for multiline return type annotations

### ISSUE THREAD

**Issue by controversial** [eslint-stylistic/eslint-stylistic#851]
### Validations

- [x] Follow our [Code of Conduct](https://github.com/eslint-stylistic/eslint-stylistic/blob/main/CODE_OF_CONDUCT.md)
- [x] Read the [Contributing Guide](https://eslint.style/contribute/guide).
- [x] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.

### Describe the bug

the return type annotations in these two cases fail `indent`:
```ts
async function foo(
  bar: number,
): Promise<
  number
> {
  return 2;
}

function foo2(
  bar: number,
): (
  number
) {
  return 2;
}
```

the linter would rather i format it like this:
```ts
async function fooFixed(
  bar: number,
): Promise<
    number
  > {
  return 2;
}

function foo2Fixed(
  bar: number,
): (
  number
  ) {
  return 2;
}
```

These seem obviously wrong to me—the closing brackets are not indented to match the opening one. Additionally, the indentation of the inner `number` is inconsistent between the angle brackets and the parentheses.

When writing the function arguments on a single line instead, the behavior changes. The following (correctly) passes `indent` even though this indentation failed when the arguments were multiline:
```ts
async function foo4(bar: number): Promise<
  number
> {
  return 2;
}
```

however the following fails and wants `number` to have no indentation before it.
```
function foo3(bar: number): (
  number
) {
  return 2;
}
```

---

The `indent` behavior is inconsistent between parentheses vs type parameters here, and it’s wrong in all cases when a multiline return type annotation follows multiline arguments.

### Reproduction

https://stackblitz.com/edit/github-nfm2nump?file=src%2Findex.tsx

### Contributes

- [ ] I am willing to submit a PR to fix this issue
- [x] I am willing to submit a PR with failing tests

**Comment by controversial** [comment]
This bug can also be triggered for arrow functions, but it’s a bit more convoluted because the return type annotation must include a closing parenthesis token.

The following return type annotation fails `indent`:

```ts
const a = (
  param: 2 | 3,
): Promise<
  (2 | 3)
> => {
  return Promise.resolve(param)
}
```

without the parentheses around the `2 | 3` inside Promise<> it works fine.
